lightning/src/events/bump_transaction.rs

```diff
@@ -0,0 +1,597 @@
+		tx_lock_time: PackedLockTime,
+	},
+}
+
+pub struct Input {
+	pub outpoint: OutPoint,
+	pub witness_weight: u64,
+}
+
+#[derive(Clone)]
+pub struct Utxo {
+	pub outpoint: OutPoint,
+	pub output: TxOut,
+	pub witness_weight: u64,
+}
+
+pub struct CoinSelection {
+	confirmed_utxos: Vec<Utxo>,
+	fee_amount_sat: u64,
+	additional_tx_outputs: Vec<TxOut>,
+}
+
+pub trait CoinSelectionSource {
+	fn select_confirmed_utxos(
+		&self, must_spend: Vec<Input>, must_pay_to: Vec<TxOut>, target_feerate_sat_per_1000_weight: u32,
+	) -> Result<CoinSelection, ()>;
+	fn change_script(&self) -> Result<Script, ()>;
+	fn sign_confirmed_utxo(&self, tx: &Transaction, input: usize) -> Result<Witness, ()>;
+}
+
+pub trait WalletSource {
+	fn list_confirmed_utxos(&self) -> Result<Vec<Utxo>, ()>;
+	fn change_script(&self) -> Result<Script, ()>;
+	fn sign_confirmed_utxo(&self, tx: &Transaction, input: usize) -> Result<Witness, ()>;
+}
+
+fn select_confirmed_utxos<W: Deref>(
+	wallet: &W, target_feerate_sat_per_1000_weight: u32, target_amount: Option<u64>,
+	preexisting_tx_weight: u64,
+) -> Result<CoinSelection, ()> where W::Target: WalletSource {
+	let mut utxos = wallet.list_confirmed_utxos()?;
+	utxos.sort_unstable_by_key(|utxo| utxo.output.value);
+
+	let change_script = wallet.change_script()?;
+	let target_amount = target_amount.unwrap_or_else(|| change_script.dust_value().to_sat());
+
+	// Mostly copied from BDK's `LargestFirstCoinSelection`.
+	// https://github.com/bitcoindevkit/bdk/blob/master/src/wallet/coin_selection.rs#L210
+	let mut selected_amount = 0;
+	let mut fee_amount = preexisting_tx_weight * target_feerate_sat_per_1000_weight as u64;
+	let selected = utxos.into_iter().scan(
+		(&mut selected_amount, &mut fee_amount), |(selected_amount, fee_amount), utxo| {
+			if **selected_amount < target_amount + **fee_amount {
+				**fee_amount += target_feerate_sat_per_1000_weight as u64 * (41 * 4 + utxo.witness_weight);
+				**selected_amount += utxo.output.value;
+				Some(utxo)
+			} else {
+				None
+			}
+		}
+		).collect::<Vec<_>>();
+
+	let amount_needed_with_fees = target_amount + fee_amount;
+	if selected_amount < amount_needed_with_fees {
+		return Err(());
+	}
+
+	let remaining_amount = selected_amount - amount_needed_with_fees;
+	let change_output_fee = target_feerate_sat_per_1000_weight as u64
+		* (8 + change_script.consensus_encode(&mut sink()).map_err(|_| ())? as u64);
+	let change_output_amount = remaining_amount.saturating_sub(change_output_fee);
+	let change_output = if change_output_amount < change_script.dust_value().to_sat() {
+		None
+	} else {
+		Some(TxOut { script_pubkey: change_script, value: change_output_amount })
+	};
+
+	Ok(CoinSelection {
+		confirmed_utxos: selected,
+		fee_amount_sat: fee_amount,
+		additional_tx_outputs: change_output.map(|co| vec![co]).unwrap_or(Vec::new()),
+	})
+}
+
+pub enum UtxoSource<C: Deref, W: Deref>
+where
+	C::Target: CoinSelectionSource,
+	W::Target: WalletSource,
+{
+	CoinSelection(C),
+	Wallet(W),
```

Comment:
Haven't thought it through a ton to see if it would work, but would it simplify things a bit to provide a default implementation of `CoinSelectionSource` that would take in a `WalletSource` from the user, and then you wouldn't have to differentiate between the variants?